### ISSUE TITLE
Give name to a rule (and may be an id??)

### ISSUE THREAD

**Issue by arsalans** [CacheControl/json-rules-engine#25]
@CacheControl  Is there a way we can give names to our rules? So for example:

  ```
let volumeRule = new Rule({
    **name: myVolumeRule  //----- Note the name here** 
    conditions: {
      all: [{
        fact: 'volume',
        operator: 'greaterThanInclusive',
        value: 100
      }]
    },
    event: {  // define the event to fire when the conditions evaluate truthy
      type: 'fouledOut',
      params: {
        message: 'Player has fouled out!'
      }
    }
    ,
    priority: 100
  });
```

So the reason for this is that when you are writing 100s of rules, you may want to give them name to identify which rule you are talking about. Also, if we are saving them into a database, we can save them in rows and identified by their names. Its helpful when you are managing a large number of rules.

**Comment by arsalans** [comment]
Do note however that if I do specify a name, it does take it and nothing bad happens. However, i am also assuming that the engine is not checking for duplicate names either.

**Comment by slmmm** [comment]
@arsalans I'm guessing that you might be wanting to name the rule so that you are able to identify which rule triggered (emits event).  What I did was carry the name of my rule in the "params" property object, along with any other meta-data type of information I needed if a rule triggered.   You're right, the engine won't mind extra properties, but it doesn't carry them across.

**Comment by CacheControl** [comment]
This is a good feedback.

@arsalans - What are the specific use case(s) where you need the custom properties to be attached to the rule object?  If you just want to tack on a property to a rule instance, you could do so after invoking the constructor:
```js
let volumeRule = new Rule({ .. })
volumeRule.name = 'VolumeGreaterThan50'
```

Is the crux of the issue that the `name` property isn't represented when the rule is converted to JSON?  Or is it the issue that slmmm brought up?  Or perhaps both? 😄 

@slmmm To your workaround:  If we were to address the above issue of custom properties, what do you think about the rule being provided as the third argument to the event emission, e.g.:
```js
engine.on('success', (event, almanac, rule) => {
  console.log(rule.name) //  'VolumeGreaterThan50'
})
```

**Comment by arsalans** [comment]
@CacheControl I haven't put the rules in database yet but I will be doing this activity soon. So yes, I will need it then to differentiate the rules.

Also, engine should check if you have two rules with the same name and error or warn about this condition.

I like your suggestion regarding passing the whole rule object for every event, it does help in debugging your rules.

**Comment by slmmm** [comment]
@CacheControl In the past, I thought about your suggestion (i.e. the rule on event) as a what-if and it seemed like an easy win until I imagined the likelihood of using the different properties of the rule.  Realistically, I don't imagine that "conditions" would be very useful after event emission because that's already been utilized for the purpose it was intended (by the engine). So after "conditions", what's left? The remainder of the rule properties are lean and realistically, I could carry around attributes I was interested in on the "params" as easily as on the "rule" (tomatoe, tomato).   My data service layer gets/sets the JSON structure for the rule as it's needed for the rule engine versus how it's represented in saving to a backend store.  Maybe providing the rule would give a better opportunity to structure the rule with properties that are directly attached to the rule itself rather than hidden in an "unrelated" property called "params" (aka parameters) - playing with more/less intuitive description of rule JSON itself. 

Do you see some other gains I might have missed?

**Comment by CacheControl** [comment]
@slmmm Nope, you're 100% correct!  This project is coming up on a year old, and I think I'm starting to forget why certain design decisions were made.  Thanks for keeping me honest!  Also, I'd like to take this opportunity to thank you for your input on this project - it is very much appreciated.

@arsalans I've persisted rules in a database with this project before, so perhaps I can provide some guidance.  Assuming you're using a relational database, I think it's best to think about each rule as being a _column_, rather than a _row_.   So after calling `rule.toJSON()`, you might take that value and store it in a `rule_definition` column.  Our table schema looked something like the following:

```sql
CREATE TABLE `rules` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(256) NOT NULL,
  `rule_definition` TEXT NOT NULL,  # this is where rule.toJSON() would go
  `create_time` datetime NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `ux_name` (`name`)
) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8;
```

This allows you to add things like foreign keys to other resources (is the rule "owned" by something?), and provide columns for other metadata you find useful.

If you're going the NoSQL approach, you'd probably want to follow the same conceptual pattern; simply nest the `rule.toJSON()` results into a dedicated property of whatever resource you're persisting.

Finally, I see the responsibility of enforcing things like rule uniqueness on being outside the scope of this library - as modeled above, this is something most database technologies (or if working on the front end, libraries like lodash) do very well already, and I'd prefer to keep `json-rules-engine` lightweight and un-opinionated in this regard; a thing that runs rules and spits out results.

Does this address your concerns?

**Comment by murraybauer** [comment]
My two cents (and thanks for a great library):

1) I think any data structures for persistence should be outside the rule engine/data structure used for execution context.  I persist my rules with a wrapper data structure in a NoSQL store similiar to @CacheControl that includes name, description, disabled/enabled states and other common fields like timestamp, created_by, modified etc. Everyone's will be different - it has no bearing to the rules engine itself.

2) Also, re duplicate name validation - why would the execution engine be concerned with duplicate names - it's not, it shouldn't alter it's behavior. It's job is to execute all the rules passed in.  Once again domain specific logic.  Quite easy to filter out beforehand, in my case that includes disabled rules.

3) Passing in the rule as a paramater - this is a must for me.  My use case is that user's create/dynamically build business rules via UI e.g. Volume->GreaterThan->50.  When the rule passes or fails, I need to be able to print out a log message saying 'Rule failed - the volume was 30, and must be greater than 50'.  Right now I had not way to access the 'rule' object without duplicating the entire conditions object into event params - ironically as part of my persistence logic!

**Comment by CacheControl** [comment]
@murraybauer thanks for presenting your scenario, this is helpful.  I'd like to dig in a bit more so we solve this for everyone's possible use case:

If your goal is to log an appropriate message on `success`, and you had access to the `rule` object, how would you know which condition(s) succeeded if you were dealing with an **any** operator?  Similarly, with `failure` + **all**, you might have X conditions succeed and only one fail, so it'd be impossible to know what to log, even with access to the rule object.

This sounds like the desire to know the result of each condition that was run.  e.g. if you had a `ANY(Volume->GreaterThan->50, Height->LessThan->10`) rule, you'd want to know that the Height condition succeeded, but the Volume condition failed, so you can print the appropriate message of `Rule failed - the volume was 30, and must be greater than 50`.

Thoughts?

**Comment by murraybauer** [comment]
@CacheControl you raise a very good point!   I am chasing exactly the logging you have outlined.

To explain to previous comment, I have two implementations of the rules engine (one as a large generic rules engine TBD) and one as a simple means to make a some user application settings more flexible.  In this case, I actually only care about success and it's single row of AND conditions (like an advanced tag picker of sorts). e.g If A,B,C.. are all true, take this action, so the hack works in this small case.

The other related use case for needing to access a fact's value from the event in addition to logging, is mainly because the action/event I need to take (in my case if the entire rule passes) requires access to the fact value(s) by factName. e.g. `increase the Width of the cube by [50%] of the Height`.  For static facts that I pass into the engine, I can in theory get the fact value I need from the same source, but dynamic facts that are executed only when needed and compute values (expensive/async logic) is more problematic as I then would need to do things like cache the last result etc etc

I have to do some more testing on implementing this, but it seems you can get a fact's value using:

A) `engine.getFact('factName')` - but maybe not for dynamic async facts.
B) or `almanac.factValue('accountId').then()` as per example 04-fact-dependancy. almanac seams to be exposed on `engine.on('success', (event, almanac) => {` but not `engine.run().then(events)`

**Comment by CacheControl** [comment]
hey @murraybauer I've implemented the ability to determine individual condition success and failure, published **2.0.0-beta1**.  There's an [example](https://github.com/CacheControl/json-rules-engine/blob/7974907ba41a78a8ac6a195229ea04c28b8fd2de/examples/09-rule-results.js) in pr #35 and some docs [here](https://github.com/CacheControl/json-rules-engine/blob/7974907ba41a78a8ac6a195229ea04c28b8fd2de/docs/rules.md#rule-results) and [here](https://github.com/CacheControl/json-rules-engine/blob/7974907ba41a78a8ac6a195229ea04c28b8fd2de/docs/rules.md#events)

This a beta because I'm going to let it stew for a week or so to provide the various stakeholders an opportunity to give feedback.  Please let me know if you have any input.

**Comment by murraybauer** [comment]
thanks @CacheControl appreciate the heads up and suburb work on rule results!  The example looks to be a perfect fit.  Will implement beta2 to test in the next few days and let you know.

**Comment by CacheControl** [comment]
Closing - feature requests have been addressed in 2.0.0
